README.md
# Entity Query Parser

This is the entity query language parser for TypeScript.

The project contains a lexer and a parser which both are based on a
continunation passing style approach. State is represented as a
function that is called with a context, which then returns the next
state function while mutating the context.

The output is a list of EQU operators and operands in reverse polish
notation which will be transformed to queries by other projects.

## License

See LICENSE file.

## Copyright

homefully GmbH


CHANGELOG.md
# v0.5.40 (Wed Oct 13 2021)

#### 🐛 Bug Fix

- Update dependency typescript to v4.4.4 [#122](https://github.com/homefully/equ-parser/pull/122) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.39 (Tue Oct 05 2021)

#### 🐛 Bug Fix

- Update dependency auto to v10.32.1 [#118](https://github.com/homefully/equ-parser/pull/118) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.38 (Tue Sep 21 2021)

#### 🐛 Bug Fix

- Update minor-upgrades [#115](https://github.com/homefully/equ-parser/pull/115) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.37 (Tue Sep 14 2021)

#### 🐛 Bug Fix

- Update minor-upgrades [#112](https://github.com/homefully/equ-parser/pull/112) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.36 (Tue Aug 31 2021)

#### 🐛 Bug Fix

- Update dependency typescript to v4.4.2 [#109](https://github.com/homefully/equ-parser/pull/109) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.35 (Tue Aug 17 2021)

#### 🐛 Bug Fix

- Update dependency auto to v10.31.0 [#107](https://github.com/homefully/equ-parser/pull/107) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.34 (Tue Jul 27 2021)

#### 🐛 Bug Fix

- Update dependency auto to v10.30.0 [#102](https://github.com/homefully/equ-parser/pull/102) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.33 (Wed Jul 07 2021)

#### 🐛 Bug Fix

- Update dependency @types/jest to v26.0.24 [#101](https://github.com/homefully/equ-parser/pull/101) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.32 (Tue Jul 06 2021)

#### 🐛 Bug Fix

- Update dependency typescript to v4.3.5 [#99](https://github.com/homefully/equ-parser/pull/99) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.31 (Tue Jun 29 2021)

#### 🐛 Bug Fix

- Update dependency prettier to v2.3.2 [#98](https://github.com/homefully/equ-parser/pull/98) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.30 (Tue Jun 22 2021)

#### 🐛 Bug Fix

- Update dependency typescript to v4.3.4 [#96](https://github.com/homefully/equ-parser/pull/96) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.29 (Tue Jun 08 2021)

#### 🐛 Bug Fix

- Update dependency auto to v10.29.3 [#95](https://github.com/homefully/equ-parser/pull/95) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.28 (Sat Jun 05 2021)

#### 🐛 Bug Fix

- Update minor-upgrades [#93](https://github.com/homefully/equ-parser/pull/93) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.27 (Tue May 25 2021)

#### 🐛 Bug Fix

- Update dependency auto to v10.29.2 [#90](https://github.com/homefully/equ-parser/pull/90) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.26 (Tue May 25 2021)

#### ⚠️ Pushed to `master`

- Update dependency auto to v10.29.0 ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.25 (Tue May 18 2021)

#### 🐛 Bug Fix

- Update dependency auto to v10.27.1 [#85](https://github.com/homefully/equ-parser/pull/85) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.24 (Tue May 11 2021)

#### 🐛 Bug Fix

- Update minor-upgrades [#84](https://github.com/homefully/equ-parser/pull/84) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.23 (Tue May 04 2021)

#### 🐛 Bug Fix

- Update dependency auto to v10.26.1 [#83](https://github.com/homefully/equ-parser/pull/83) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.22 (Tue Apr 27 2021)

#### 🐛 Bug Fix

- Update dependency @types/jest to v26.0.23 [#82](https://github.com/homefully/equ-parser/pull/82) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.21 (Tue Apr 20 2021)

#### 🐛 Bug Fix

- Update minor-upgrades [#79](https://github.com/homefully/equ-parser/pull/79) ([@renovate-bot](https://github.com/renovate-bot))

#### Authors: 1

- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.5.20 (Tue Apr 20 2021)

#### 🐛 Bug Fix

- Update dependency auto to v10 [#81](https://github.com/homefully/equ-parser/pull/81) ([@renovate-bot](https://github.com/renovate-bot) [@renovate[bot]](https://github.com/renovate[bot]))

#### Authors: 2

- [@renovate[bot]](https://github.com/renovate[bot])

test/equ.test.ts
import { parse } from "../src/equ"

const example = `name.first[eq:"foo"|eq:"bar"]|email[ct:"foo",ct:"bar"],age[(gt:4.5,lt:-10)|eq:15]`

describe("equ", () => {
    it("should parse a string to its RPN representation", () => {
        const result = parse(example)

        expect(result).toMatchSnapshot()
    })

    it("should parse exist operator with booleans", () => {
        const result = parse("path[ex:true],path2[ex:false]")

        expect(result).toMatchSnapshot()
    })

    it("should throw an error on parsing ex operator without a boolean", () => {
        expect(() => parse("path[ex:1]")).toThrowErrorMatchingSnapshot()
    })

    it('should throw an error on missing "', () => {
        expect(() => parse("path[eq:TEST2]")).toThrowErrorMatchingSnapshot()
        expect(() => parse('path[eq:"TEST2]')).toThrowErrorMatchingSnapshot()
    })
})


test/Dummy.spec.ts
import { test } from "../src/Dummy"

describe("Dummy", () => {
    it("should return \"test\"", () => {
        expect(test()).toBe("test")
    })
})


test/lexer.test.ts
import { lex } from "../src/lexer"

const example = `name.first[eq:"foo"|eq:"bar"]|email[ct:"foo",ct:"bar"],age[(gt:4.5,lt:-10)|eq:15]`

describe("lexer", () => {
    it("should lex the example correctly", () => {
        const result = lex(example)

        expect(result).toMatchSnapshot()
    })

    it("should throw an EQUParsingError on error", () => {
        expect(() => lex("ioi[")).toThrowErrorMatchingSnapshot()
    })

    it("should throw a parsing error on unended string values", () => {
        expect(() => lex(`path[eq:"asd`)).toThrowErrorMatchingSnapshot()
    })

    it("should lex booleans correctly", () => {
        const result = lex(`path[ex:true],path2[ex:false]`)

        expect(result).toMatchSnapshot()
    })

    it("should lex bundles correctly", () => {
        const result = lex(`(path[eq:"asd"],path2[eq:"bdf"])`)

        expect(result).toMatchSnapshot()
    })

    it("should lex expression bundles correctly", () => {
        const result = lex(`path[eq:"asd"|gt:"asdf",lt:"asdf"]`)

        expect(result).toMatchSnapshot()
    })

    it("should throw an error if the path is missing", () => {
        expect(() => lex("[")).toThrowErrorMatchingSnapshot()
    })

    it("should throw an error on missing start filters", () => {
        expect(() => lex("as")).toThrowErrorMatchingSnapshot()
    })

    it("should throw an error on missing filterIs", () => {
        expect(() => lex("as[eq]")).toThrowErrorMatchingSnapshot()
    })

    it("should throw an error on an invalid filter value", () => {
        expect(() => lex("as[eq:[]")).toThrowErrorMatchingSnapshot()
    })

    it("should correctly parse esacpe sequences", () => {
        const result = lex(`path[eq:"string with \\\\ in it and also \\""]`)

        expect(result).toMatchSnapshot()
    })

    it("should throw an error on an invalid escape sequence", () => {
        expect(() => lex(`path[eq:"\\s"]`)).toThrowErrorMatchingSnapshot()
    })

    it("should throw an error on an invalid number", () => {
        expect(() => lex("path[eq:.12.12]")).toThrowErrorMatchingSnapshot()
    })

    it("should throw an error on invalid combinator", () => {
        expect(() =>
            lex("path[eq:123]:path[eq:123]")
        ).toThrowErrorMatchingSnapshot()
    })

    it("should lex decimal numbers correctly", () => {
        const result = lex("path[eq:-123.123]")
        const result2 = lex("path[eq:.123]")

        expect(result).toMatchSnapshot()
        expect(result2).toMatchSnapshot()
    })

    it("should lex a date correctly", () => {
        const result = lex("path[eq:2020-01-01]")

        expect(result).toMatchSnapshot()
    })

    it("should lex zulu date time", () => {
        const result = lex("path[eq:2020-01-01T12:00:00.000Z]")

        expect(result).toMatchSnapshot()
    })

    it("should lex offset date time", () => {
        const result = lex("path[eq:2020-01-01T12:00:00.000+01:00]")

        expect(result).toMatchSnapshot()
    })

    it("should lex offset date time without colon", () => {
        const result = lex("path[eq:2020-01-01T12:00:00.000+0100]")

        expect(result).toMatchSnapshot()
    })

    it("should not lex dates with invalid formats", () => {
        expect(() => lex("path[eq:2020-1-1]")).toThrowErrorMatchingSnapshot()
    })

    it("should not lex incomplete dates", () => {
        expect(() => lex("path[eq:2020-23]")).toThrowErrorMatchingSnapshot()
    })

    it("should not lex incomplete date time values", () => {
        expect(() =>
            lex("path[eq:2020-10-10T12]")
        ).toThrowErrorMatchingSnapshot()
        expect(() =>
            lex("path[eq:2020-10-10T12:12]")
        ).toThrowErrorMatchingSnapshot()
        expect(() =>
            lex("path[eq:2020-10-10T12:12:12]")
        ).toThrowErrorMatchingSnapshot()
        expect(() =>
            lex("path[eq:2020-10-10T12:12:12.123]")
        ).toThrowErrorMatchingSnapshot()
        expect(() =>
            lex("path[eq:2020-10-10T12:12:12.123+]")
        ).toThrowErrorMatchingSnapshot()
        expect(() =>
            lex("path[eq:2020-10-10T12:12:12.123+12]")
        ).toThrowErrorMatchingSnapshot()
        expect(() =>
            lex("path[eq:2020-10-10T12:12:12.123+12:]")
        ).toThrowErrorMatchingSnapshot()
    })
})


src/parser.ts
import { Item, filters } from "./lexer"
import { EQUParsingError } from "./error"

function expressionPrecedence(
    a: ExpressionOperator,
    b: ExpressionOperator
): boolean {
    const list = ["or", "and", "bundleStart", "bundleEnd"]

    return list.indexOf(a.operator) > list.indexOf(b.operator)
}

function filterPrecedence(a: FilterOperator, b: FilterOperator): boolean {
    const list = ["or", "and", "bundleStart", "bundleEnd"]

    return list.indexOf(a.operator) > list.indexOf(b.operator)
}

function parseError(msg?: any): never {
    if (typeof msg === "string") {
        throw new EQUParsingError(msg)
    }

    throw new EQUParsingError("unexpected " + JSON.stringify(msg))
}

type StateFn = ((ctx: ParserContext) => StateFn) | undefined

type FilterOperatorType = "and" | "or" | "bundleStart" | "bundleEnd"

type ExpressionType = keyof typeof filters
type ExpressionOperatorType = "and" | "or" | "bundleStart" | "bundleEnd"

type ValueType = "number" | "string" | "boolean" | "date" | "dateTime"

export type FilterOperator = {
    type: "operator"
    operator: FilterOperatorType
}

export type FilterOperand = {
    type: "operand"
    path: string
    expressions: Array<ExpressionItem>
}

export type ParseItem = FilterOperand | FilterOperator

export type ExpressionOperand = {
    type: "expressionOperand"
    expressionType: ExpressionType
    valueType: ValueType
    value: string | number | boolean | Date
}

export type ExpressionOperator = {
    type: "expressionOperator"
    operator: ExpressionOperatorType
}

export type ExpressionItem = ExpressionOperator | ExpressionOperand

class ParserContext {
    items: Array<Item>
    parsedItems: Array<ParseItem> = []

    filterOperatorStack: Array<FilterOperator> = []
    expressionOperatorStack: Array<ExpressionOperator> = []

    constructor(items: Array<Item>) {
        this.items = items
    }

    next(): Item {
        return this.items.shift() ?? parseError("expected to have more items")
    }

    pushFilterItem(item: ParseItem) {
        this.parsedItems.push(item)
    }

    pushExpressionItem(item: ExpressionItem) {
        const current = this.currentFilterOperand()

        current.expressions.push(item)
    }

    popFilterOperator(): FilterOperator | undefined {
        return this.filterOperatorStack.pop()
    }

    popExpressionOperator(): ExpressionOperator | undefined {
        return this.expressionOperatorStack.pop()
    }

    topExpressionOperator(): ExpressionOperator | undefined {
        if (this.expressionOperatorStack.length === 0) {
            return undefined
        }

        return this.expressionOperatorStack[
            this.expressionOperatorStack.length - 1
        ]
    }

    topFilterOperator(): FilterOperator | undefined {
        if (this.filterOperatorStack.length === 0) {
            return undefined
        }

        return this.filterOperatorStack[this.filterOperatorStack.length - 1]
    }

    currentFilterOperand(): FilterOperand {
        for (let i = this.parsedItems.length - 1; i >= 0; i--) {
            const item = this.parsedItems[i]
            if (item.type === "operand") {
                return item
            }
        }

        // this code is effectively unreachable. this only gets called when the
        // state is setup such that there is a FilterOperand in the output
        /* istanbul ignore next */
        parseError("expected to have a filter operand, but didn't find one")
    }
}

function parsePath(ctx: ParserContext): StateFn {
    const next = ctx.next()

    if (next.type === "bundleStart") {
        return parseFilterBundleStart
    }

    if (next.type === "path") {
        ctx.pushFilterItem({
            type: "operand",
            path: next.str,
            expressions: [],
        })

        return parseFiltersStart
    }

    parseError(next)
}

function parseFilterBundleStart(ctx: ParserContext): StateFn {
    ctx.filterOperatorStack.push({
        type: "operator",
        operator: "bundleStart",
    })

    return parsePath
}

function parseFiltersStart(ctx: ParserContext): StateFn {
    const next = ctx.next()
    if (next.type === "filtersStart") {
        return parseExpression
    }

    parseError(next)
}

function parseExpression(ctx: ParserContext): StateFn {
    const expressionType = ctx.next()

    if (expressionType.type === "bundleStart") {
        return parseExpressionBundleStart
    }

    let expressionOperator: ExpressionType = "eq"
    switch (expressionType.type) {
        case "filterCt": {
            expressionOperator = "ct"
            break
        }
        case "filterEq": {
            expressionOperator = "eq"
            break
        }
        case "filterGt": {
            expressionOperator = "gt"
            break
        }
        case "filterGte": {
            expressionOperator = "gte"
            break
        }
        case "filterLt": {
            expressionOperator = "lt"
            break
        }
        case "filterLte": {
            expressionOperator = "lte"
            break
        }
        case "filterRgx": {
            expressionOperator = "rgx"
            break
        }
        case "filterEx": {
            expressionOperator = "ex"
            break
        }
        case "filterNeq": {
            expressionOperator = "neq"
            break
        }
        default:
            parseError(expressionType)
    }

    const value = ctx.next()
    let valueType: ValueType = "string"
    switch (value.type) {
        case "string": {
            valueType = "string"
            break
        }
        case "number": {
            valueType = "number"
            break
        }
        case "boolean": {
            valueType = "boolean"
            break
        }
        case "date": {
            valueType = "date"
            break
        }
        case "dateTime": {
            valueType = "dateTime"
            break
        }
        default:
            parseError(value)
    }

    if (expressionOperator === "ex" && valueType !== "boolean") {
        parseError(
            "encountered ex operator with the value " +
                value.str +
                ", but it must be boolean"
        )
    }

    let parsedValue: number | string | boolean | Date = value.str
    if (value.type === "number") {
        parsedValue = Number(value.str)
    }
    if (value.type === "boolean") {
        switch (value.str) {
            case "true": {
                parsedValue = true
                break
            }
            case "false": {
                parsedValue = false
                break
            }
            default:
                parseError("invalid boolean: " + value.str)
        }
    }

    if (value.type === "date" || value.type === "dateTime") {
        parsedValue = new Date(value.str)
        if (isNaN(parsedValue.getTime())) {
            parseError("invalid date value: " + value.str)
        }
    }

    ctx.pushExpressionItem({
        type: "expressionOperand",
        expressionType: expressionOperator,
        valueType,
        value: parsedValue,
    })

    return parseAfterExpression
}

function parseExpressionBundleStart(ctx: ParserContext): StateFn {
    ctx.expressionOperatorStack.push({
        type: "expressionOperator",
        operator: "bundleStart",
    })

    return parseExpression
}

function parseAfterExpression(ctx: ParserContext): StateFn {
    const next = ctx.next()

    if (next.type === "filtersEnd") {
        const currentFilterOperand = ctx.currentFilterOperand()
        currentFilterOperand.expressions.push(
            ...ctx.expressionOperatorStack.reverse()
        )
        ctx.expressionOperatorStack = []

        return parseAfterFilter
    }

    if (next.type === "or" || next.type === "and") {
        const expressionOperator: ExpressionOperator = {
            type: "expressionOperator",
            operator: next.type,
        }

        let currentTopExpression = ctx.topExpressionOperator()
        while (
            currentTopExpression !== undefined &&
            expressionPrecedence(expressionOperator, currentTopExpression)
        ) {
            const top = ctx.popExpressionOperator() ?? parseError()
            ctx.pushExpressionItem(top)
            currentTopExpression = ctx.topExpressionOperator()
        }

        ctx.expressionOperatorStack.push(expressionOperator)

        return parseExpression
    }

    if (next.type === "bundleEnd") {
        let op = ctx.popExpressionOperator()
        while (op !== undefined) {
            if (op.operator === "bundleStart") {
                break
            }

            ctx.pushExpressionItem(op)
            op = ctx.popExpressionOperator()
        }
        return parseAfterExpression
    }

    parseError(next)
}

function parseAfterFilter(ctx: ParserContext): StateFn {
    const next = ctx.next()

    if (next.type === "eof") {
        ctx.parsedItems.push(...ctx.filterOperatorStack.reverse())
        ctx.filterOperatorStack = []

        return undefined
    }

    if (next.type === "and" || next.type === "or") {
        const filterOperator: FilterOperator = {
            type: "operator",
            operator: next.type,
        }

        let currentTopFilter = ctx.topFilterOperator()
        while (
            currentTopFilter !== undefined &&
            filterPrecedence(filterOperator, currentTopFilter)
        ) {
            const top = ctx.popFilterOperator() ?? parseError()
            ctx.pushFilterItem(top)
            currentTopFilter = ctx.topFilterOperator()
        }

        ctx.filterOperatorStack.push(filterOperator)

        return parsePath
    }

    if (next.type === "bundleEnd") {
        let op = ctx.popFilterOperator()
        while (op !== undefined) {
            if (op.operator === "bundleStart") {
                break
            }

            ctx.pushFilterItem(op)
            op = ctx.popFilterOperator()
        }
        return parseAfterFilter
    }

    parseError(next)
}

export function parse(items: Array<Item>): Array<ParseItem> {
    const parser = new ParserContext(items)
    let state: StateFn = parsePath
    while (state !== undefined) {
        state = state(parser)
    }

    return parser.parsedItems
}


src/lexer.ts
import assert from "assert"
import { EQULexingError } from "./error"

export type Token =
    | "eof"
    | "error"
    | "path"
    | "or"
    | "and"
    | "filtersStart"
    | "filtersEnd"
    | "bundleStart"
    | "bundleEnd"
    | "string"
    | "number"
    | "boolean"
    | "date"
    | "dateTime"
    | "filterGt"
    | "filterLt"
    | "filterGte"
    | "filterLte"
    | "filterEq"
    | "filterCt"
    | "filterRgx"
    | "filterEx"
    | "filterNeq"

export const filters: { [k: string]: Token } = {
    eq: "filterEq",
    lt: "filterLt",
    gt: "filterGt",
    lte: "filterLte",
    gte: "filterGte",
    ct: "filterCt",
    rgx: "filterRgx",
    ex: "filterEx",
    neq: "filterNeq",
}

const tokenBundleStart = "("
const tokenBundleEnd = ")"

const tokenOr = "|"
const tokenAnd = ","

const tokenFiltersStart = "["
const tokenFiltersEnd = "]"

const tokenFilterIs = ":"

const tokenValidPath =
    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789._"
const tokenValidOperator = "abcdefghijklmnopqrstuvwxyz"
const tokenValidNumberStart = "+-.0123456789"

const tokenValidDigits = "0123456789"

const tokenTrue = "true"
const tokenFalse = "false"

type StateFn = ((ctx: LexerContext) => StateFn) | undefined

function lexError(msg?: string): never {
    throw new EQULexingError(msg)
}

class LexerContext {
    input: string
    pos: number = 0
    items: Array<Item> = []
    start: number = 0
    width: number = 0

    constructor(str: string) {
        this.input = str
    }

    emit(token: Token) {
        let value = this.input.substring(this.start, this.pos)

        if (token === "string") {
            value = value.replace("\\\\", "\\")
            value = value.replace('\\"', '"')
        }

        const item = { str: value, type: token }

        this.items.push(item)
        this.start = this.pos
    }

    next(length?: number): string {
        const next = this.input.substr(this.pos, length ?? 1)
        this.width = next.length
        this.pos += this.width
        return next
    }

    accept(valid: string): boolean {
        if (valid.indexOf(this.next()) >= 0) {
            return true
        }

        this.backup()
        return false
    }

    backup() {
        this.pos -= this.width
    }

    peek(length?: number): string {
        const next = this.next(length)
        this.backup()
        return next
    }

    acceptRun(valid: string) {
        while (true) {
            const next = this.next()
            if (next === "") {
                break
            }

            if (valid.indexOf(next) === -1) {
                break
            }
        }

        this.backup()
    }

    acceptExact(str: string): boolean {
        const match = this.input.substring(this.pos).startsWith(str)
        if (match) {
            this.pos += str.length
            this.width = str.length
        }
        return match
    }

    acceptRunUntil(...stops: Array<string>) {
        while (true) {
            const next = this.next()
            if (next === "") {
                this.backup()
                return false
            }

            if (stops.indexOf(next) >= 0) {
                this.backup()
                return true
            }
        }
    }

    ignore() {
        this.start = this.pos
    }

    rest(): string {
        return this.input.substring(this.start)
    }

    current(): string {
        return this.input.substring(this.start, this.pos)
    }

    assertLength(length: number, error: string) {
        if (this.pos - this.start !== length) {
            lexError(error)
        }
    }

    assertLengthBetween(min: number, max: number, error: string) {
        const length = this.pos - this.start
        if (!(min <= length && length <= max)) {
            lexError(error)
        }
    }
}

export type Item = {
    str: string
    type: Token
}

function lexFilters(ctx: LexerContext): StateFn {
    if (ctx.peek() === tokenBundleStart) {
        return lexToken(tokenBundleStart, "bundleStart", lexFilters)
    }

    ctx.acceptRun(tokenValidPath)
    if (ctx.pos <= ctx.start) {
        lexError("expected path, but got " + ctx.rest())
    }

    ctx.emit("path")

    if (ctx.peek() !== tokenFiltersStart) {
        lexError("expected '[', but got " + ctx.rest())
    }

    return lexToken(tokenFiltersStart, "filtersStart", lexInsideFilters)
}

function lexInsideFilters(ctx: LexerContext): StateFn {
    if (ctx.peek() === tokenBundleStart) {
        return lexToken(tokenBundleStart, "bundleStart", lexInsideFilters)
    }

    ctx.acceptRun(tokenValidOperator)
    const nextFilter = ctx.current()
    const filterType =
        filters[nextFilter] ??
        lexError("expected filter type, got " + nextFilter)
    ctx.emit(filterType)

    if (ctx.next() !== tokenFilterIs) {
        lexError("expected " + tokenFilterIs + " but got " + ctx.current())
    }

    ctx.ignore()

    return lexFilterValue
}

function lexFilterValue(ctx: LexerContext): StateFn {
    const next = ctx.peek()

    if (ctx.acceptExact(tokenTrue) || ctx.acceptExact(tokenFalse)) {
        ctx.emit("boolean")
        return lexAfterFilterValue
    }

    if (tokenValidNumberStart.indexOf(next) >= 0) {
        return lexNumberOrDateTime
    }

    if (next === '"') {
        ctx.next()
        ctx.ignore()
        return lexStringValue
    }

    lexError("expected filter value but got " + ctx.rest())
}

function lexNumberOrDateTime(ctx: LexerContext): StateFn {
    const peek = ctx.peek()

    if (peek === "+" || peek === "-" || peek === ".") {
        return lexNumberValue
    }

    ctx.acceptRun(tokenValidDigits)
    if (ctx.accept("-")) {
        ctx.assertLength(5, "invalid date value: " + ctx.current())
        return lexRestOfDateTime
    }

    ctx.accept(".")
    ctx.acceptRun(tokenValidDigits)
    ctx.emit("number")

    return lexAfterFilterValue
}

function lexRestOfDateTime(ctx: LexerContext): StateFn {
    // month
    ctx.acceptRun(tokenValidDigits)
    ctx.assertLength(7, "invalid date value: " + ctx.current())

    if (!ctx.accept("-")) {
        lexError("invalid date value: " + ctx.current())
    }

    // day
    ctx.acceptRun(tokenValidDigits)
    ctx.assertLength(10, "invalid date value: " + ctx.current())

    if (ctx.accept("T")) {
        return lexRestOfTime
    }

    ctx.emit("date")
    return lexAfterFilterValue
}

function lexRestOfTime(ctx: LexerContext): StateFn {
    // hour
    ctx.acceptRun(tokenValidDigits)
    ctx.assertLength(
        13,
        "invalid date time value: " +
            ctx.current() +
            "; hour is not 2 digits long"
    )
    if (!ctx.accept(":")) {
        lexError("invalid date time value: " + ctx.current())
    }

    // minute
    ctx.acceptRun(tokenValidDigits)
    ctx.assertLength(
        16,
        "invalid date time value: " +
            ctx.current() +
            "; minute is not 2 digits long"
    )
    if (!ctx.accept(":")) {
        lexError("invalid date time value: " + ctx.current())
    }

    // second
    ctx.acceptRun(tokenValidDigits)
    ctx.assertLength(
        19,
        "invalid date value: " + ctx.current() + "; second is not 2 digits long"
    )
    if (!ctx.accept(".")) {
        lexError("invalid date time value: " + ctx.current())
    }

    // second fraction
    ctx.acceptRun(tokenValidDigits)
    ctx.assertLength(
        23,
        "invalid date value: " +
            ctx.current() +
            "; second fraction is not 3 digits long"
    )
    if (ctx.accept("Z")) {
        ctx.emit("dateTime")
        return lexAfterFilterValue
    }

    if (!ctx.accept("-+")) {
        lexError("invalid date time value: " + ctx.current())
    }

    // offset hour
    ctx.acceptRun(tokenValidDigits)
    ctx.accept(":")

    // offset minute
    ctx.acceptRun(tokenValidDigits)
    ctx.assertLengthBetween(28, 29, "invalid date time value: " + ctx.current())
    ctx.emit("dateTime")

    return lexAfterFilterValue
}

function lexNumberValue(ctx: LexerContext): StateFn {
    ctx.accept("+-")

    let decimalOccured = ctx.accept(".")

    ctx.acceptRun(tokenValidDigits)
    if (ctx.accept(".")) {
        if (decimalOccured) {
            lexError("invalid number: " + ctx.current())
        }

        ctx.acceptRun(tokenValidDigits)
    }

    ctx.emit("number")
    return lexAfterFilterValue
}

function lexStringValue(ctx: LexerContext): StateFn {
    while (true) {
        const found = ctx.acceptRunUntil("\\", '"')
        if (!found) {
            lexError("expected string to end, but got " + ctx.rest())
        }

        const next = ctx.peek()
        if (next === '"') {
            ctx.emit("string")

            ctx.next()
            ctx.ignore()
            return lexAfterFilterValue
        }

        // if we're here next must be a \ because of acceptRunUntil
        assert(next === "\\")
        ctx.next()

        const escapedSymbol = ctx.peek()

        if (escapedSymbol === "\\" || escapedSymbol === '"') {
            ctx.next()
        } else {
            lexError("invalid escape sequence: \\" + escapedSymbol)
        }
    }
}

function lexAfterFilterValue(ctx: LexerContext): StateFn {
    const next = ctx.next()
    switch (next) {
        case tokenOr: {
            ctx.emit("or")
            break
        }
        case tokenAnd: {
            ctx.emit("and")
            break
        }
        case tokenFiltersEnd: {
            ctx.backup()
            return lexToken(tokenFiltersEnd, "filtersEnd", lexAfterFilter)
        }
        case tokenBundleEnd: {
            ctx.backup()
            return lexToken(tokenBundleEnd, "bundleEnd", lexAfterFilterValue)
        }
    }

    return lexInsideFilters
}

function lexAfterFilter(ctx: LexerContext): StateFn {
    const next = ctx.next()

    if (next === "") {
        ctx.emit("eof")
        return undefined
    }

    switch (next) {
        case tokenOr: {
            ctx.emit("or")
            return lexFilters
        }
        case tokenAnd: {
            ctx.emit("and")
            return lexFilters
        }
        case tokenBundleEnd: {
            ctx.backup()
            return lexToken(tokenBundleEnd, "bundleEnd", lexAfterFilter)
        }
    }

    lexError("invalid combinator: " + next)
}

function lexToken(tokenValue: string, token: Token, next: StateFn): StateFn {
    return (ctx: LexerContext) => {
        ctx.pos += tokenValue.length
        ctx.emit(token)
        return next
    }
}

export function lex(str: string): Array<Item> {
    const lexer = new LexerContext(str)

    let state: StateFn = lexFilters
    while (state !== undefined) {
        state = state(lexer)
    }

    return lexer.items
}
